pkg/util/json/json.go

```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
 	return jsonObject(append([]jsonKeyValuePair(nil), b.pairs...)), nil
 }
 
+// PartialObject is a JSON object builder that builds an object using a set of
+// fixed key-values (supplied at construction) and a set of keys which are
+// settable when building the final JSON object.
+type PartialObject struct {
+	basePairs      []jsonKeyValuePair
+	keyOrd         map[string]int
+	allowedNewKeys []string
+}
+
+// NewPartialObject creates a PartialObject with the specified base key-value
+// pairs, and allows for setting the values of newKeys when building the final
+// JSON object.
+func NewPartialObject(base map[string]JSON, newKeys []string) (*PartialObject, error) {
+	// Check that newKeys are unique.
+	newKeySet := make(map[string]struct{}, len(newKeys))
+	for _, k := range newKeys {
+		if _, ok := newKeySet[k]; ok {
+			return nil, errors.AssertionFailedf("expected unique keys, found %v", newKeys)
+		}
+		newKeySet[k] = struct{}{}
+	}
+
+	sort.Strings(newKeys)
```

Comment:
This sort isn't necessary given we're sorting again below after appending `newKeys`